fix: enhance AzureOpenAIResponsesAPIConfig to support different Azure auth method like we have for completions api.

hsuyuming opened the pull request:
Auth method like we have for completions api.

## Enhance AzureOpenAIResponsesAPIConfig to support different Azure auth method 
support  Entra ID for Azure Auth /  Azure Username and Password for Azure Auth /  Azure OIDC Token / Azure AD token provider based on Service Principal with Secret workflow for Azure Auth


## Relevant issues

Reference to issue https://github.com/BerriAI/litellm/issues/10868

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [✓] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [✓] I have added a screenshot of my new test passing locally 
- [✓] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [✓] My PR's scope is as isolated as possible, it only solves 1 specific problem



## Test reuslt
```

## Result of make test-unit 
Those 3 error is not coming from my change
```
(.venv) user@abehsu-us-vscode-med:~/abehsu/litellm_clone$ make test-unit
poetry run pytest tests/litellm/
============================================ test session starts ============================================
platform linux -- Python 3.10.12, pytest-7.4.4, pluggy-1.5.0
rootdir: /home/user/abehsu/litellm_clone
plugins: requests-mock-1.12.1, asyncio-0.21.2, anyio-4.5.2, respx-0.22.0, mock-3.14.0
asyncio: mode=strict
collected 800 items / 3 errors                                                                              

================================================== ERRORS ===================================================
____________________ ERROR collecting tests/litellm/enterprise/test_enterprise_routes.py ____________________
ImportError while importing test module '/home/user/abehsu/litellm_clone/tests/litellm/enterprise/test_enterprise_routes.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/litellm/enterprise/test_enterprise_routes.py:13: in <module>
    from litellm_enterprise.proxy.enterprise_routes import router
E   ModuleNotFoundError: No module named 'litellm_enterprise.proxy'
__________________ ERROR collecting tests/litellm/litellm_core_utils/test_token_counter.py __________________
ImportError while importing test module '/home/user/abehsu/litellm_clone/tests/litellm/litellm_core_utils/test_token_counter.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/litellm/litellm_core_utils/test_token_counter.py:26: in <module>
    from tests.large_text import text
E   ModuleNotFoundError: No module named 'tests.large_text'
_______________ ERROR collecting tests/litellm/litellm_core_utils/test_token_counter_tool.py ________________
ImportError while importing test module '/home/user/abehsu/litellm_clone/tests/litellm/litellm_core_utils/test_token_counter_tool.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/litellm/litellm_core_utils/test_token_counter_tool.py:13: in <module>
    from test_token_counter import token_counter
tests/litellm/litellm_core_utils/test_token_counter.py:26: in <module>
    from tests.large_text import text
E   ModuleNotFoundError: No module named 'tests.large_text'
============================================= warnings summary ==============================================
tests/litellm/integrations/test_custom_prompt_management.py:27
  /home/user/abehsu/litellm_clone/tests/litellm/integrations/test_custom_prompt_management.py:27: PytestCollectionWarning: cannot collect test class 'TestCustomPromptManagement' because it has a __init__ constructor (from: tests/litellm/integrations/test_custom_prompt_management.py)
    class TestCustomPromptManagement(CustomPromptManagement):

tests/litellm/integrations/arize/test_arize_utils.py:181
  /home/user/abehsu/litellm_clone/tests/litellm/integrations/arize/test_arize_utils.py:181: PytestCollectionWarning: cannot collect test class 'TestArizeLogger' because it has a __init__ constructor (from: tests/litellm/integrations/arize/test_arize_utils.py)
    class TestArizeLogger(CustomLogger):

tests/litellm/litellm_core_utils/test_streaming_handler.py:505
  /home/user/abehsu/litellm_clone/tests/litellm/litellm_core_utils/test_streaming_handler.py:505: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=3)

tests/litellm/llms/ollama/test_ollama_chat_transformation.py:16
  /home/user/abehsu/litellm_clone/tests/litellm/llms/ollama/test_ollama_chat_transformation.py:16: PytestCollectionWarning: cannot collect test class 'TestEvent' because it has a __init__ constructor (from: tests/litellm/llms/ollama/test_ollama_chat_transformation.py)
    class TestEvent(BaseModel):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========================================== short test summary info ==========================================
ERROR tests/litellm/enterprise/test_enterprise_routes.py
ERROR tests/litellm/litellm_core_utils/test_token_counter.py
ERROR tests/litellm/litellm_core_utils/test_token_counter_tool.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================= 4 warnings, 3 errors in 4.24s =======================================
make: *** [Makefile:29: test-unit] Error 2
```


## Type

🐛 Bug Fix

## Changes

hsuyuming committed:
fix: enhance AzureOpenAIResponsesAPIConfig to support different Azure Auth method like we have for completions api.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=10871) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=10871) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=10871) it.</sub>

hsuyuming commented:
@ishaan-jaff  i check with the pipeline error, it is cause by module note find, how can i fix this issue?
```text
tests/litellm/enterprise/test_enterprise_routes.py:13: in <module>
    from litellm_enterprise.proxy.enterprise_routes import router
E   ModuleNotFoundError: No module named 'litellm_enterprise.proxy'
```

ishaan-jaff commented on litellm/llms/azure/responses/transformation.py:
```diff
@@ -18,16 +24,73 @@
+        litellm_params: dict,
         api_key: Optional[str] = None,
     ) -> dict:
+        azure_ad_token_provider = litellm_params.get("azure_ad_token_provider")
```
can we make this new block of logic a simple helper in azure/common_utils.py ? 

Then we can re-use for new endpoints like /image/edits etc

ishaan-jaff replied:
I mean the entire new section of code you added, that should be a simple helper util

ishaan-jaff reviewed (changes requested):
LGTM, just 1 change

hsuyuming committed:
extract duplicate code to common functoin as util for future reference

hsuyuming commented:
@ishaan-jaff  I update my code base on your suggestion, please help me review again when you get time. Thank you!

hsuyuming committed:
remove unused import

hsuyuming committed:
Add type annotation in order to fix this issue.
llms/azure/common_utils.py:416: error: Incompatible types in assignment (expression has type "AsyncClient | None", target has type "str | Callable[[], str] | None")  [assignment]
llms/azure/common_utils.py:418: error: Incompatible types in assignment (expression has type "Client | None", target has type "str | Callable[[], str] | None")  [assignment]

hsuyuming commented:
fix all of CI issues

krrishdholakia commented:
@ishaan-jaff is this okay to merge?

ishaan-jaff reviewed (changes requested):
1 change, mostly looks good

hsuyuming commented on litellm/llms/azure/common_utils.py:
```diff
@@ -259,6 +259,67 @@ def select_azure_base_url_or_endpoint(azure_client_params: dict):
+            azure_ad_token = get_azure_ad_token_provider()
+        except ValueError:
+            verbose_logger.debug("Azure AD Token Provider could not be used.")
+    return api_key, azure_ad_token_provider, azure_ad_token
```
i will update the code within this week. Thank you for the feedback!

hsuyuming committed:
convert response to pydantic  object and also provide unittest for it

hsuyuming committed:
remove unused type

ishaan-jaff reviewed (approved):
LGTM !

hsuyuming commented:
@ishaan-jaff  I create a pull request to your litellm_staging_azure_responses_api branch.
I fetch the latest code from origin, then do the rebase. hope this step is correct. If not, feel free to let me know which command i should use. Thank you
https://github.com/BerriAI/litellm/pull/11035
